browser/src/libs/code_intelligence/code_intelligence.tsx

```diff
@@ -259,6 +259,29 @@ export interface CodeHost extends ApplyLinkPreviewOptions {
     codeViewsRequireTokenization?: boolean
 }
 
+export interface BaseFileDiff {
+    diffType: string
+}
+export type FileDiff<T extends FileInfo = FileInfo> = AddedFileDiff<T> | ModifiedFileDiff<T> | RemovedFileDiff<T>
+export interface AddedFileDiff<T extends FileInfo = FileInfo> extends FileDiffWithHead<T> {
+    diffType: 'added'
+}
+export interface ModifiedFileDiff<T extends FileInfo = FileInfo> extends FileDiffWithHead<T>, FileDiffWithBase<T> {
+    diffType: 'modified'
+}
+export interface RemovedFileDiff<T extends FileInfo = FileInfo> extends FileDiffWithBase<T> {
+    diffType: 'removed'
+}
+export type FileDiffWithContent = FileDiff<FileInfoWithContent>
+export type FileDiffWithRepoNames = FileDiff<FileInfoWithRepoNames>
+
+export interface FileDiffWithHead<T extends FileInfo = FileInfo> extends BaseFileDiff {
+    head: T
+}
+export interface FileDiffWithBase<T extends FileInfo = FileInfo> extends BaseFileDiff {
+    base: T
+}
```

Comment:
(ignore if this is wip and you were already planning on changing diff, just a suggestion) for conciseness, I would declare `FileDiff` as a single union type, rather than split it into multiple interfaces.

```suggestion
export type FileDiff<T extends FileInfo = FileInfo> = { diffType: 'added'; head: T } | { diffType: 'removed'; base: T } | { diffType: 'modified'; head: T; base: T }
```

Reply:
I simplified the type definitions somewhat, but I kept the diff interfaces for now and I commented below about where they seemed useful